### PR TITLE
udhcpc: no MSG_DONTROUTE when sending packet

### PR DESCRIPTION
**Pull request by adiov**
**TL;DR** Don't set `MSG_DONTROUTE` when sending packets, otherwise it forces the client into non-compliant behaviour that is rejected by properly configured servers, thus creating unnecessary interruptions and connection breakages.

This patch reverts a [change made in Sep 2017 which introduced `MSG_DONTROUTE` flag](http://lists.busybox.net/pipermail/busybox-cvs/2017-September/037402.html) to prevent udhcpc from reaching out to servers on a different subnet. That change violates [RFC2131 ](https://tools.ietf.org/html/rfc2131)by forcing fully configured clients, who got their configurations through an offer relayed by a DHCP relay, from renewing through a unicast request directly to the DHCP server, resulting in the client resorting to boradcasting lease extension requests instead of unicasting them, further breaking [RFC2131](https://tools.ietf.org/html/rfc2131).

The problem with `MSG_DONTROUTE` appears when talking to a properly configured DHCP server that rejects non-compliant requests. Such server will reject lease extension attempts sent via broadcast rather than unicast, as is the case with Finnish ISPs Telia and DNA as well as Estonian ISP Starman. Once the lease expires without renewal, udhcpc enters init mode, taking down the interfaces with it, and thus causing interruption on every lease expiry. On some ISPs (such as the ones mentioned above) that can be once every 10-20 minutes. The interruptions appear in the logs as such:

```
udhcpc: sending renew to x.x.x.x
udhcpc: send: Network unreachable
udhcpc: sending renew to 0.0.0.0
udhcpc: sending renew to 0.0.0.0
...
udhcpc: lease lost, entering init state
Interface 'wan' has lost the connection
Interface 'wan' is now down
Network alias 'eth0' link is down
udhcpc: sending select for y.y.y.y
udhcpc: lease of y.y.y.y obtained, lease time 1200
Network alias 'eth0' link is up
Interface 'wan' is now up
```
This keeps repeating every lease period, breaking VPN connections, SSH sessions, etc.

During lease extension, a fully configured client should be able to [reach out to the server from which it received the lease for extension](https://www.netmanias.com/en/post/techdocs/6000/dhcp-network-protocol/understanding-dhcp-relay-agents), regardless in which network it is; that's up to the gateway to find. This patch ensures that.

It's worth noting that this correct behaviour exists in every mainstream DHCP client I checked, including the [Internet System Consortium's](https://fossies.org/linux/dhcp/client/dhclient.c).

**P.S.** I've made the change locally, built it, and have been testing it on my Archer C7 v2 for quite a while. This part of the logs from half an hour ago shows it's working as expected without interruption:
```
Wed Jun  6 22:39:07 2018 daemon.notice netifd: wan (1299): udhcpc: sending renew to x.x.x.x
Wed Jun  6 22:39:07 2018 daemon.notice netifd: wan (1299): udhcpc: lease of y.y.y.y obtained, lease time 1200
Wed Jun  6 22:49:07 2018 daemon.notice netifd: wan (1299): udhcpc: sending renew to x.x.x.x
Wed Jun  6 22:49:07 2018 daemon.notice netifd: wan (1299): udhcpc: lease of y.y.y.y obtained, lease time 1200
Wed Jun  6 22:59:07 2018 daemon.notice netifd: wan (1299): udhcpc: sending renew to x.x.x.x
Wed Jun  6 22:59:07 2018 daemon.notice netifd: wan (1299): udhcpc: lease of y.y.y.y obtained, lease time 1200
```

### CODE REVIEW

**Comment by jow-** [comment]
Impressive work on tracking this down! LGTM. Did you send this upstream by any chance? I think the busybox authors should be made aware of this problem.

cc @dedeckeh

**Comment by dedeckeh** [comment]
@adiov Indeed nice work tracking this down; we can accept the patch but it would be ideal upstream busybox fixes this as well.
If I understand it correctly the issue arises when a DHCP relay is present in the network; did you observe the giaddr field being set in the DHCP offer/ack coming from the DHCP server ? If yes this could be used by the client as an indication it has no direct connection with the DHCP server and as result don't set the MSG_DONROUTE flag.

**Comment by adiov** [comment]
This morning, I filed a [bug and attached the patch in BusyBox's Bugzilla](https://bugs.busybox.net/show_bug.cgi?id=11081). Let's see how it that progresses.

In the dumps I collected, `GIADDR` is being set by the relay in both `DHCPOFFER` and `DHCPACK`  messages.

image

Having that said, I don't believe `MSG_DONROUTE` should be set even if `GIADDR` is not observed, for two reasons:

1. A misbehaving actor capable of injecting a bogus `Server Identifier` can also inject or omit `GIADDR`, give a false `SIADDR`, and pretty much manipulate the whole DHCP exchange to achieve whatever purpose they have in mind. In my opinion, conditionally setting `MSG_DONROUTE` is as useful as not setting it at all.
2. No other DHCP client is doing it.

I suggest that no flag is passed.

**Comment by ldir-EDB0** [comment]
Had this patch in local tree for 24 hours - All is well for me, in fact DHCPv4 renewals with my ISP are cleaner, going straight to the DHCP server.  Before this change the direct route would fail and would fall back to a broadcast renewal.

Tested-by: Kevin Darbyshire-Bryant <ldir@darbyshire-bryant.me.uk>

**Comment by dedeckeh** [comment]
As the issue is critical I've pushed the patch to both master and openwrt-18.06. If upstream busybox fixes the issue in another way we can backport the patch; thx

**Comment by adiov** [comment]
Tested with default `OpenWrt SNAPSHOT r7137-e4259bed3f` on TP-Link Archer C7 v2. It works as expected.

```
Fri Jun  8 09:50:48 2018 daemon.notice netifd: wan (1322): udhcpc: sending renew to x.x.x.x
Fri Jun  8 09:50:48 2018 daemon.notice netifd: wan (1322): udhcpc: lease of y.y.y.y obtained, lease time 1200
```
`x.x.x.x` is on a different network.

**Comment by adiov** [comment]
@dedeckeh BusyBox upstream fixed the issue
https://bugs.busybox.net/show_bug.cgi?id=11081#c3
https://git.busybox.net/busybox/commit/?id=148788eb0ee96026105755cf3fd1ad3d94f49cd2

**Comment by dedeckeh** [comment]
@adiov Thanks for informing me; replaced the patch by the upstream "fix"
